query-engine/core/src/query_document/mod.rs

```diff
@@ -60,30 +64,74 @@ impl BatchDocument {
+            // Otherwise, we just look for a scalar field inside the model. If it's not one, then we break.
+            val => match model.fields().find_from_scalar(&key) {
+                Ok(_) => match val {
+                    // Consider scalar _only_ if the filter object contains "equals". eg: `{ scalar_field: { equals: 1 } }`
+                    QueryValue::Object(obj) => !obj.contains_key(filters::EQUALS),
+                    _ => false,
+                },
+                Err(_) => true,
+            },
+        })
+    }
+
+    fn can_compact(&self, schema: &QuerySchemaRef) -> bool {
         match self {
             Self::Multi(operations, _) => match operations.split_first() {
-                Some((first, rest)) if first.is_find_unique() => rest.iter().all(|op| {
-                    op.is_find_unique()
-                        && first.name() == op.name()
-                        && first.nested_selections().len() == op.nested_selections().len()
-                        && first
-                            .nested_selections()
-                            .iter()
-                            .all(|fop| op.nested_selections().contains(fop))
-                }),
+                Some((first, rest)) if first.is_find_unique() => {
+                    let has_identical_selection_sets = rest.iter().all(|op| {
+                        op.is_find_unique()
+                            && first.name() == op.name()
+                            && first.nested_selections().len() == op.nested_selections().len()
+                            && first
+                                .nested_selections()
+                                .iter()
+                                .all(|fop| op.nested_selections().contains(fop))
+                    });
+
+                    // If any of the operation has an "invalid" filter (see documentation of `operation_contain_invalid_filter`),
+                    // we do not compact the queries.
+                    let has_invalid_filters = operations
+                        .iter()
+                        .any(|op| Self::operation_contain_invalid_filter(op, schema));
+
+                    has_identical_selection_sets && !has_invalid_filters
+                }
                 _ => false,
             },
             Self::Compact(_) => false,
         }
     }
 
-    pub fn compact(self) -> Self {
+    pub fn compact(self, schema: &QuerySchemaRef) -> Self {
         match self {
-            Self::Multi(operations, _) if self.can_compact() => Self::Compact(CompactedDocument::from(operations)),
+            Self::Multi(operations, _) if self.can_compact(schema) => {
+                Self::Compact(CompactedDocument::from_operations(operations, schema))
+            }
             _ => self,
         }
     }
+
+    /// Returns `true` if the batch document is [`Compact`].
+    #[must_use]
+    pub fn is_compact(&self) -> bool {
+        matches!(self, Self::Compact(..))
+    }
```

Comment:
Out of curiosity, nothing actionable.

As far as I can see, this function is used in tests. Is there a reason why we prefer to define these type of helpers in the code rather than tests, when they use public modules like `BatchDocument::Compact`?

Reply:
Tbh, there's no real reason. It's true that I could've used `assert!(matches!(doc, BatchDocument::Compact(_)))`. On the other hand, I don't feel like leaking this kind of this API should ever hurt us. If you think otherwise, please let me know